### PR TITLE
#304 New records guide

### PR DESCRIPTION
**Pull request by akiroz**
connects #304

### CODE REVIEW

**Comment by oursky-travis** [comment]
@akiroz, thanks for your PR! By analyzing the annotation information on this pull request, we identified @TakLee96, @rickmak and @ben181231 to be potential reviewers

**Comment by akiroz** [comment]
@bensonby 
The spell checker is having problems with UUIDs and names... should I add them to the spelling file?
screenshot from 2016-10-20 13-59-40

**Comment by bensonby** [comment]
@akiroz yes, there are some exceptions under the section "html code block" in the `.spelling` file. You can rename the section name, so it contains all exceptions about the code we used

File: .spelling [hunk 1]
```diff
@@ -125,6 +125,7 @@ utils
 versioning
 whitelist
 writable
+sniplet
```
Comment (bensonby) [inline .spelling]: should be "snippet", no need to insert here

File: content/js/record.md [hunk 2]
```diff
@@ -1,335 +1,189 @@
...
-tables in SQL.
-
-``` javascript
-const Note = skygear.Record.extend('note');
-const Blog = skygear.Record.extend('blog');
-
-const note = new Note({ 'content': 'Hello World' });
+<a name="what-are-records"></a>
+## What are Records?
+In Skygear, you can store data using **Records**. They provide a reliable way store *structured*, *persistent* data on the backend.
```
Comment (bensonby) [inline content/js/record.md]: beware of the 80-character line limit as the convention

File: content/js/record.md [hunk 3]
```diff
@@ -1,335 +1,189 @@
...
+
+Each Skygear record is associated with a named **type** which is bound to an extendable data schema. In the JS SDK, records are simply JS objects.
+
+All Skygear records are also assigned a unique UUID during creation for identification.
```
Comment (bensonby) [inline content/js/record.md]: can have an example to illustrate what a Record is? I am not sure if a reader can understand by reading these paragraph?

File: content/js/record.md [hunk 4]
```diff
@@ -1,335 +1,189 @@
...
+
+<a name="record-owner"></a>
+### Record Owner
+By design, each Skygear record is associated a user for easy access control so you must be logged-in to be able to create and save records. However, if access control is not required, you can login as the anonymous user.
```
Comment (bensonby) [inline content/js/record.md]: - Record owner is not mentioned in this paragraph, need to mention that the "user" is the record owner, and that the record owner can read/edit the record
- Need a link to Access Control page

File: content/js/record.md [hunk 5]
```diff
@@ -1,335 +1,189 @@
...
+
+<a name="record-owner"></a>
+### Record Owner
+By design, each Skygear record is associated a user for easy access control so you must be logged-in to be able to create and save records. However, if access control is not required, you can login as the anonymous user. 
+
+Refer to the [User & Authentication][] section for more detail.
+
+<a name="record-object"></a>
+### Record Object
+Skygear records are JS objects, you can easily create records using the following 2 lines:
```
Comment (bensonby) [inline content/js/record.md]: - need to separate the 2 js lines with the bob object as separate code blocks
- Only `recordType` and `id` are assigned upon record creation

File: content/js/record.md [hunk 6]
```diff
@@ -1,335 +1,189 @@
...
-To control the access, you may set different access control to the record.
-- The database objects can be accessed with `skygear.publicDB` and
-`skygear.privateDB`.
+Refer to the [Access Control][] section for more detail.
 
-<a name="basic-crud"></a>
-## Basic CRUD
+<a name="sql-database-backend"></a>
+### SQL Database Backend
+On the Skygear server backend, a relational SQL database is used to store your data. The record's type name is mapped to a table name while field names are mapped to table columns. You can view and modify skygear's data by connecting directly to the database.
```
Comment (bensonby) [inline content/js/record.md]: can mention that it is PostgreSQL

File: content/js/record.md [hunk 7]
```diff
@@ -1,335 +1,189 @@
...
-})).then((record) => {
-  console.log(record);
-}, (error) => {
-  console.error(error);
-});
+<a name="save"></a>
+## Save
+You can save skygear records using a single API call, here's how you could save the `bob` record created in the section [above][record object]:
+```js
+const savePromise = skygear.publicDB.save(bob);
```
Comment (bensonby) [inline content/js/record.md]: Involving only 2 lines, we can add how to create `bob` here.

File: content/js/record.md [hunk 8]
```diff
@@ -1,335 +1,189 @@
...
 ```
+`savePromise` is an [A+ Promise Object][] that either returns the record that was successfully saved or the error when the save operation failed. You can find out more about errors in the [Error Handling][] guide.
```
Comment (bensonby) [inline content/js/record.md]: - should be "`save` returns an.."?
- needs to mention the `save` function for `privateDB`
- the sample code needs to show the simplest form of error handling for easy understanding. It doesn't matter if we have it in every piece of codes

File: content/js/record.md [hunk 9]
```diff
@@ -1,335 +1,189 @@
...
-query.equalTo('_owner', skygear.currentUser.id);
-// '_owner' is an alias for '_owner_id'
-skygear.publicDB.query(query);
+<a name="atomic-save-updates"></a>
+## Atomic Save/Updates
+In a multi-save/update operation, you can make it such that either all or none of the records will be
+saved if an error occurs using the `atomic` option. This call will behave just like a single-item save
+operation:
+```js
+skygear.publicDB.save(arrayOfRecords, { atomic: true });
```
Comment (bensonby) [inline content/js/record.md]: Needs example on success/error handling to show how it behaves like a single-item save

**Comment by bensonby** [comment]
@rickmak you can review the technical parts of this doc

File: content/js/record.md (hunk 5 above)
Comment (akiroz) [inline content/js/record.md]: I checked, all are assigned upon creation.

File: .spelling (hunk 1 above)
Comment (akiroz) [inline .spelling]: aspell in vim said it was a word LOL

**Comment by rickmak** [comment]
Look good to me.

**Comment by bensonby** [comment]
I am creating a new branch for this, because some content (data types, reference, etc.) will be removed after merging this PR. @boris you can submit PR about data type later on this branch: `js-record-data-type-restructure`.

**Comment by akiroz** [comment]
ok~

**Comment by bensonby** [comment]
Picked as a5122f7. Will need grammar and styling review, issue created in #379

**Pull request closed without merging**